server/src/main/java/org/elasticsearch/discovery/PeerFinder.java

```diff
@@ -129,11 +132,13 @@ public void activate(final DiscoveryNodes lastAcceptedNodes) {
 
     public void deactivate(DiscoveryNode leader) {
         final boolean peersRemoved;
-        final List<Releasable> connectionReferences;
+        final Collection<Releasable> connectionReferences = new ArrayList<>();
```

Comment:
The default size is probably more than enough in most cases, but maybe worth a comment? Also we allocate a similarly-sized `Set` under the mutex lower down, so maybe we can do a right-sized allocation for this collection under the mutex instead?